x-pack/filebeat/docs/inputs/input-websocket.asciidoc

```diff
@@ -0,0 +1,275 @@
+[role="xpack"]
+
+:type: websocket
+:mito_version: v1.8.0
+:mito_docs: https://pkg.go.dev/github.com/elastic/mito@{mito_version}
+
+[id="{beatname_lc}-input-{type}"]
+=== Websocket Input
+experimental[]
+
+The `websocket` input reads messages from a websocket server or api endpoint. This input uses the `CEL engine` and the `mito` library interally to parse and process the messages. Having support for `CEL` allows you to parse and process the messages in a more flexible way. It has many similarities with the `cel` input as to how the `CEL` programs are written but deviates in the way the messages are read and processed. The `websocket` input is a `streaming` input and can only be used to read messages from a websocket server or api endpoint. 
+
+This input supports:
+
+* Auth
+** Basic
+** Bearer 
+** Custom
+
+NOTE: The `websocket` input as of now does not support XML messages. Auto-reconnects are also not supported at the moment so reconnection will occur on input restart.
+
+==== Execution
+
+The execution environment provided for the input includes includes the functions, macros, and global variables provided by the mito library.
+A single JSON object is provided as an input accessible through a `state` variable. `state` contains a `response` map field and may contain arbitrary other fields configured via the input's `state` configuration. If the CEL program saves cursor states between executions of the program, the configured `state.cursor` value will be replaced by the saved cursor prior to execution.
+
+On start the `state` will be something like this:
+
+["source","json",subs="attributes"]
+----
+{
+    "response": { ... },
+    "cursor": { ... },
+    ...
+}
+----
+The `websocket` input creates a `response` field in the state map and attaches the websocket message to this field. All `CEL` programs written should act on this `response` field. Additional fields may be present at the root of the object and if the program tolerates it, the cursor value may be absent.Only the cursor is persisted over restarts, but all fields in state are retained between iterations of the processing loop except for the produced events array, see below.
+
+If the cursor is present the program should perform and process requests based on its value.
+If cursor is not present the program must have alternative logic to determine what requests to make.
+
+After completion of a program's execution it should return a single object with a structure looking like this:
+
+["source","json",subs="attributes"]
+----
+{
+    "events": [ <1>
+        {...},
+        ...
+    ],
+    "cursor": [ <2>
+        {...},
+        ...
+    ],
```

Comment:
Let's have the cursor be present for now since the input atm is being released as experimental. Based on the real world behaviour and use cases we can later remove it if it's not useful.